### DISCUSSION TITLE
General Project Discussion

### DISCUSSION THREAD

**Discussion by jeffjohannsen** [NBA-Betting/NBA_Betting#43] [General]
Creating this discussion as the main location for questions, ideas, and other areas of interest about this project, NBA Betting algorithms, and the NBA in general.

**Comment by jeffjohannsen** [comment]
Will also be posting any updates about the project that aren't adequately explained in the Readme.

**Comment by jeffjohannsen** [comment]
### Outline of Version 2

1. **Player Prediction Framework** - Combination of ML/DL and Traditional Algorithms used to predict the performance of individual players.
2. **Player Interaction and Synergy Framework** - Focused on understanding and predicting the performance of players in various combinations on and off the court.
Example: How does Steph Curry perform when Draymond Green is on the court vs off the court?
3. **Team Prediction Framework** - Combination of ML/DL and traditional algorithms used to predict the performance of teams as a whole.
4. **Game Prediction Framework** - Betting market data and other misc predictors of game outcomes.
5. **Prediction Framework Integration** - Combining the 4 prediction frameworks into a single bet management system. Final testing, organization, and cleanup. Full deployment to the web app.

- I have migrated the project management from GitHub Projects to Jira. Issues will re-generate from there when I start on them.

- Version 2 is now the focus of the project. This will take over the main branch in a couple of months when all Version 1 work is superseded by the Team Prediction Framework of Version 2.

**Reply by aaaclones** [reply]
for problem 2 . I find it more effective if you just can put a utility value on a player when hes on the floor vs. off the floor , I am not talking about +/- either , perhaps some sort of advanced stat of some sort .. For baseball this stat is called (WAR) wins above replacement . Synergy framework applies when the starting lineup doesn't deviate but as we found out first hand through the season players are constantly in and out of the lineup and this creates a linear effect across the board. Outside the top 3 players on the team i am not sure how important the rest of the players are

**Reply by jeffjohannsen** [reply]
> I find it more effective if you just can put a utility value on a player when hes on the floor vs. off the floor , I am not talking about +/- either , perhaps some sort of advanced stat of some sort .. For baseball this stat is called (WAR) wins above replacement

I am currently thinking of this problem as 2 parts. One part is focused on the on/off data like you mention. Probably will use a combination of multiple advanced stats. A lot of the premier NBA advanced stats are based on +/- of some sort like EPM, AuPM, etc. but I will definitely use more than just the base +/-

**Reply by jeffjohannsen** [reply]
> Synergy framework applies when the starting lineup doesn't deviate but as we found out first hand through the season players are constantly in and out of the lineup and this creates a linear effect across the board. Outside the top 3 players on the team i am not sure how important the rest of the players are

The main part of problem 2 is determining how players being in and out of the lineup affect each others' performance. It isn't so much the synergy of a full starting lineup but of various player combinations. Using lineup combination and WOWY (with or without you) data we should be able to determine the performance of any lineup combination compared to the baseline or alternatives.

For example I am guessing that the +/- of Draymond and Curry on the floor at the same time is better than just combining their typical overall +/- ("the whole is greater than the sum of the two parts"). This can be determined for any 2,3,4,or 5 player combinations.

This way we can have a better idea of how "good" any lineup is which allows us to take into consideration players being in and out of the lineup.

**Reply by jeffjohannsen** [reply]
Overall though I am the least confident about working through the synergy and interaction framework but am excited to explore the data and see what can be useful.

**Comment by Dimus12** [comment]
Hi Jeff, 

Would it be possible to add documentation on how to set up the necessary environment and run the model?

Thanks!

**Reply by elfrost** [reply]
would  know as well, very nice project

**Reply by jeffjohannsen** [reply]
Hopefully the recent Readme update can help. If not, please feel free to ask me about any specific parts.

**Comment by txroy001** [comment]
Hi Jeff, great stuff here! Have you tried focusing on player props? In particular points, rebounds, assists, and PRA? These markets tend to be less efficient than primary betting markets like moneyline and spreads. I think you already have the pieces for it with the Player Prediction Framework and the Player Interaction and Synergy Framework.

I'm a product designer so if you're interested in help with the web app just lmk (looks great from the screenshots btw).

**Reply by jeffjohannsen** [reply]
I agree that player props are probably a better avenue. I'm trying to refocus the project towards a broader focus that can take advantage of these newer betting markets. Hopefully some progress can be obtained. Thanks for the response!

**Comment by doki81** [comment]
Hello, can a simple mortal without great programming knowledge learn to use this code

**Reply by jeffjohannsen** [reply]
I have recently updated the Readme and website to better explain the general data flow and project structure. If you have specific questions, I am happy to answer them.

**Comment by jeffjohannsen** [comment]
# Major Project Update

The project is currently in a typical data science lifecycle phase, covering data collection, storage, ETL, model predictions, and deployment. However, due to the complexity and difficulty of the problem, this method hasn't been as successful as expected, considering the costs and complexity involved.

Moving forward, the focus will shift to developing a custom model architecture. This new direction aims to avoid the most time-consuming aspect of the project—data sourcing—while also offering a novel approach to the problem.

Significant updates to the project's Readme and the [website](https://nba-betting.us/) have been made, detailing this new methodology for predicting NBA game outcomes.

Feedback and ideas about the project's future direction are highly welcomed!

**Reply by aaaclones** [reply]
Are you above 52.5 ? What are you currently predicting at for sides. ? and for totals ? if you dont mind me asking

**Reply by jeffjohannsen** [reply]
Currently, I have only been predicting the +/- line and the point differential. Specifically, my target variable for regression is home margin or how much the home team wins by (negative if the home team loses). This is converted into a boolean classification target that is true if the home margin is greater than the home margin implied by the oddsmakers line, otherwise false.

**Reply by jeffjohannsen** [reply]
The results are right around the 52.5 break even point but not enough to really be consistently and convincingly profitable.

**Reply by jeffjohannsen** [reply]
Any thoughts on how to approach this problem effectively?

**Reply by aaaclones** [reply]
I feel whatever you decide you need to implement the players on the floor a long with the team stats . Vegas is missing the line by a pretty large margin by their standards . I think with the new rules the constant shuffling of players , the standard deviation aren't just 1 offs . So whatever you decide to do you must have a way to figure out whose playing and who isn't in real time . From that you can test you test. your model however you decide to. I never had a chance to check your work or see how it's going , We were supposed to circle back to this awhile back . IF you can provide some sort of insight perhaps i can supply you with some api keys for you to access player stats past and present if that helps , I am currently working on a project  but it's not for nba it's for all sports

**Reply by jeffjohannsen** [reply]
> I feel whatever you decide you need to implement the players on the floor a long with the team stats.

I strongly agree. Injuries and rest days are probably the most important info to include in predictions. In the past, I haven't utilized them for 2 main reasons:

1. More difficult and/or costly to get both live and historical lineup data.
2. Uncertainty of how to incorporate the lineup data into the modeling process.

I started incorporating lineup data as part of the Player Prediction Framework which creates individual player predictions and aggregates them to create a game prediction. The Player Interaction and Synergy Framework would also build on this. Unfortunately, this general setup is more complicated and time-consuming on the data collection side so I have recently abandoned it. 

I am currently in the process of creating a [new system based on play-by-play data](https://github.com/NBA-Betting/NBA_Betting#custom-model-architecture) that should inherently incorporate each team's lineup at every point throughout the game.

**Reply by aaaclones** [reply]
Most certainly +/- , player defense and offense efficiency has to be included when it comes to testing for accuracy . IMO I feel those data points should be a staple in determining a lot of outcomes . Perhaps possession and pace is another key data point i didn't see in your model . Team stats can give you a seperate testing results I've explored a lot of team stat repository on github and kaggle they are never crushing the lines they hit a plateau and that's it . Due to the lineup shifts and the constant musical chairs of starting lineups the only way your ever going to come up with something special is to come up with a formula that uses player stats and then have another model that uses team stats . From that you can look at the 2 models and have a very good idea on how the team is going to perform that night . Perhaps to start you would only focus on only 2 teams to make it easier to test and have proof of concept to see how the 2 teams you are testing are doing . I think 52.5 % accuracy is profitable but if you were betting at 52.5% accuracy the edge is way to marginal and it can be quite stressful pushing such a marginal edge . In order to feel as though your really beating it you will need to be somewhere in the 58% or better for totals o/u and , 55% or better for wins or tie against the spread .

**Comment by jeffjohannsen** [comment]
[![nba_ai_logo](https://github.com/NBA-Betting/NBA_Betting/assets/63945207/5d1732e4-e95b-4498-8fef-6c2290cec925)](https://github.com/NBA-Betting/NBA_AI)

This project is no longer being actively maintained. For a more focused and updated version, check out my new project: [**NBA AI**](https://github.com/NBA-Betting/NBA_AI)
